File: app/javascript/controllers/routine_comments_controller.js
import { Controller } from "@hotwired/stimulus"
import * as bootstrap from "bootstrap"

export default class extends Controller {
  static targets = [
    "subtitle",
    "list",
    "empty",
    "body",
    "submit"
  ]

  connect() {
    this.routineValueId = null
    this.open = this.open.bind(this)

    document.addEventListener("routine-comments:open", this.open)
  }

  disconnect() {
    document.removeEventListener("routine-comments:open", this.open)
  }

  async open(event) {
    this.routineValueId = event.detail.routineValueId
    this.subtitleTarget.textContent = this.subtitle(event.detail)
    this.bodyTarget.value = ""
    this.setLoading()

    this.modal.show()

    await this.loadComments()
  }

  async loadComments() {
    const response = await fetch(this.commentsPath, {
      headers: {
        "Accept": "application/json"
      }
    })

    if (!response.ok) {
      this.showError("Não foi possível carregar os comentários.")
      return
    }

    const data = await response.json()
    this.renderComments(data.comments)
    this.broadcastCount(data.count)
  }

  async create(event) {
    event.preventDefault()

    const body = this.bodyTarget.value.trim()
    if (body === "") return

    this.submitTarget.disabled = true

    const response = await fetch(this.commentsPath, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-CSRF-Token": this.csrfToken,
        "Accept": "application/json"
      },
      body: JSON.stringify({
        routine_comment: {
          body: body
        }
      })
    })

    this.submitTarget.disabled = false

    if (!response.ok) {
      this.showError("Não foi possível salvar o comentário.")
      return
    }

    const data = await response.json()
    this.bodyTarget.value = ""
    this.appendComment(data.comment)
    this.updateEmptyState()
    this.broadcastCount(data.count)
  }

  async destroy(event) {
    event.preventDefault()

    const commentElement = event.target.closest("[data-comment-id]")
    if (!commentElement) return

    const response = await fetch(
      `${this.commentsPath}/${commentElement.dataset.commentId}`,
      {
        method: "DELETE",
        headers: {
          "X-CSRF-Token": this.csrfToken,
          "Accept": "application/json"
        }
      }
    )

    if (!response.ok) {
      this.showError("Não foi possível remover o comentário.")
      return
    }

    const data = await response.json()
    commentElement.remove()
    this.updateEmptyState()
    this.broadcastCount(data.count)
  }

  setLoading() {
    this.listTarget.innerHTML = `
      <div class="text-muted small py-3">
        Carregando comentários...
      </div>
    `
    this.emptyTarget.classList.add("d-none")
  }

  renderComments(comments) {
    this.listTarget.innerHTML = ""

    comments.forEach((comment) => {
      this.appendComment(comment)
    })

    this.updateEmptyState()
  }

  appendComment(comment) {
    this.listTarget.appendChild(this.commentElement(comment))
  }

  commentElement(comment) {
    const wrapper = document.createElement("div")
    wrapper.className = "routine-comment"
    wrapper.dataset.commentId = comment.id

    const header = document.createElement("div")
    header.className = "routine-comment__header"

    const meta = document.createElement("div")

    const user = document.createElement("strong")
    user.textContent = comment.user_name

    const time = document.createElement("span")
    time.className = "routine-comment__time"
    time.textContent = comment.created_at

    meta.appendChild(user)
    meta.appendChild(time)
    header.appendChild(meta)

    if (comment.can_destroy) {
      const button = document.createElement("button")
      button.type = "button"
      button.className = "btn btn-sm btn-link text-danger p-0"
      button.textContent = "Remover"
      button.dataset.action = "click->routine-comments#destroy"

      header.appendChild(button)
    }

    const body = document.createElement("div")
    body.className = "routine-comment__body"
    body.textContent = comment.body

    wrapper.appendChild(header)
    wrapper.appendChild(body)

    return wrapper
  }

  updateEmptyState() {
    const hasComments = this.listTarget.children.length > 0
    this.emptyTarget.classList.toggle("d-none", hasComments)
  }

  showError(message) {
    this.listTarget.innerHTML = ""

    const alert = document.createElement("div")
    alert.className = "alert alert-danger mb-0"
    alert.textContent = message

    this.listTarget.appendChild(alert)
    this.emptyTarget.classList.add("d-none")
  }

  broadcastCount(count) {
    document.dispatchEvent(
      new CustomEvent(
        "routine-comments:count-updated",
        {
          detail: {
            routineValueId: this.routineValueId,
            count: count
          }
        }
      )
    )
  }

  subtitle(detail) {
    return [
      detail.indicatorName,
      detail.referenceDate,
      detail.value
    ].filter(Boolean).join(" - ")
  }

  get commentsPath() {
    return `/routine_values/${this.routineValueId}/routine_comments`
  }

  get csrfToken() {
    return document.querySelector("meta[name='csrf-token']").content
  }

  get modal() {
    return bootstrap.Modal.getOrCreateInstance(this.element)
  }
}

File: app/controllers/routine_comments_controller.rb
class RoutineCommentsController < ApplicationController
  before_action :authenticate_user!
  before_action :set_routine_value
  before_action :set_comment, only: :destroy

  def index
    render json: {
      comments: serialized_comments,
      count: @routine_value.routine_comments.size
    }
  end

  def create
    comment = @routine_value.routine_comments.new(comment_params)
    comment.user = current_user

    if comment.save
      render json: {
        comment: serialized_comment(comment),
        count: @routine_value.routine_comments.count
      }, status: :created
    else
      render json: {
        errors: comment.errors.full_messages
      }, status: :unprocessable_entity
    end
  end

  def destroy
    return head :forbidden unless can_destroy_comment?

    @comment.destroy!

    render json: {
      count: @routine_value.routine_comments.count
    }
  end

  private

  def set_routine_value
    @routine_value =
      RoutineValue
        .includes(routine_comments: :user)
        .find(params[:routine_value_id])
  end

  def set_comment
    @comment =
      @routine_value
        .routine_comments
        .find(params[:id])
  end

  def serialized_comments
    @routine_value
      .routine_comments
      .sort_by(&:created_at)
      .map do |comment|
        serialized_comment(comment)
      end
  end

  def serialized_comment(comment)
    {
      id: comment.id,
      body: comment.body,
      user_name: routine_comment_user_name(comment.user),
      created_at: comment.created_at.strftime("%d/%m/%Y %H:%M"),
      can_destroy: comment.user_id == current_user.id || current_user.admin?
    }
  end

  def routine_comment_user_name(user)
    return user.name if user.respond_to?(:name) && user.name.present?
    return user.email if user.email.present?

    "Usuário ##{user.id}"
  end

  def can_destroy_comment?
    @comment.user_id == current_user.id || current_user.admin?
  end

  def comment_params
    params
      .require(:routine_comment)
      .permit(:body)
  end
end
